tokio-codec/src/lines_codec.rs

```diff
@@ -41,15 +105,51 @@ impl Decoder for LinesCodec {
     type Error = io::Error;
 
     fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<String>, io::Error> {
-        if let Some(newline_offset) =
-            buf[self.next_index..].iter().position(|b| *b == b'\n')
-        {
-            let newline_index = newline_offset + self.next_index;
+        let mut trim_and_offset = None;
+        for (offset, b) in buf[self.next_index..].iter().enumerate() {
+            trim_and_offset = match (b, self.max_length) {
+                // The current character is a newline, split here.
+                (&b'\n', _) => Some((1, offset)),
+                // There's a maximum line length set, and we've reached it.
+                (_, Some(max_len))
+                    if offset.saturating_sub(self.next_index) == max_len => {
+                    // If we're at the line length limit, check if the next
+                    // character(s) is a newline before we decide to return an
+                    // error.
+                    match (buf[offset + 1], buf[offset + 2]) {
```

Comment:
This will blow up if the buffer isn't actually this long.

Reply:
Ahh that's a good point. 

I added a unit test that catches the out-of-bounds panic in cb8858853b1d71b727bdc5a6adad57fda3892135, and fixed the panic in 9b72862d519239db94d3225267c41f15da571bf0. Thanks for catching that!